### PR TITLE
feat: new actor state handle

### PR DESCRIPTION
**Pull request by icorderi**
### Motivation

We need to use the new contracts from #3658

### Proposed changes

We have options.. lets talk.

### CODE REVIEW

**Comment by anorth** [comment]
I like your proposal. It looks like it should be functionally equivalent to the externally-observable behaviour the spec is aiming for. This PR is obviously missing implementation details of how those mechanics are achieved, but if achievable 👍 .

The spec also aims to prevent message send during what would be the protected regions. I think this could be added to your proposed semantics.

File: internal/pkg/vm/testing.go [hunk 1]
```diff
@@ -213,6 +229,9 @@ func (ts testStorage) Commit(cid.Cid, cid.Cid) error {
 
 // Head returns an empty Cid to satisfy the Storage interface
 func (ts testStorage) Head() cid.Cid {
+	if ts.state == nil {
+		return cid.Undef
+	}
```
Comment (icorderi) [inline internal/pkg/vm/testing.go]: @acruikshank any idea how some tests were passing before? 
`encoding.Encode({{ nil }}.state)` was blowing up during initactor tests.

Comment (acruikshank) [reply internal/pkg/vm/testing.go]: Which tests weren't passing with what changes? I don't know how to answer this question.

File: internal/pkg/vm/internal/vmcontext/actor_state_handle.go [hunk 2]
```diff
@@ -9,62 +9,110 @@ import (
...
+	// call user code allowing mutation but not side-effects
+	h.ctx.AllowSideEffects(false)
+	out, err := f()
+	h.ctx.AllowSideEffects(true)
+
+	// forward user error
+	if err != nil {
+		// re-load state from storage
+		// Note: this will throw away any changes done to the object
+		h.get(oldcid, obj)
```
Comment (anorth) [inline internal/pkg/vm/internal/vmcontext/actor_state_handle.go]: I don't think this is correct. Prior to this function call, obj was some arbitrary object pointer, probably nil. This does not restore it to its old state, but to a different state (equiv to Readonly). A correct rollback would be something like a deepcopy, but I don't think it's worth it. I propose instead that the obj will be left as the transaction left it, but won't be written to storage or have the head updated.

File: internal/pkg/vm/internal/vmcontext/actor_state_handle.go [hunk 3]
```diff
@@ -77,35 +125,57 @@ func (h *actorStateHandle) UpdateRelease(obj interface{}) {
...
 	h.head = &newcid
+
+	// add validation
+	h.addCidValidation(obj, newcid)
+
+	return out, nil
+}
+
+func (h *actorStateHandle) addCidValidation(obj interface{}, expected cid.Cid) {
+	// checks that the cid of the obj has not changed
```
Comment (anorth) [inline internal/pkg/vm/internal/vmcontext/actor_state_handle.go]:
```suggestion
	// checks that the cid of the obj matches an expected value
```

File: internal/pkg/vm/internal/vmcontext/actor_state_handle.go [hunk 4]
```diff
@@ -77,35 +125,57 @@ func (h *actorStateHandle) UpdateRelease(obj interface{}) {
 		runtime.Abort("Storage commit error")
 	}
 
-	// make new head available to take
+	// update head
 	h.head = &newcid
+
+	// add validation
+	h.addCidValidation(obj, newcid)
```
Comment (anorth) [inline internal/pkg/vm/internal/vmcontext/actor_state_handle.go]: A list of validation fns is not the right semantics here. If a method contains two well-formed transactions in sequence then the validation fn from the first one will not pass after the second. I think we want semantics of a single expected CID which is updated with each transaction.

File: internal/pkg/vm/internal/vmcontext/actor_state_handle.go [hunk 5]
```diff
@@ -77,35 +125,57 @@ func (h *actorStateHandle) UpdateRelease(obj interface{}) {
...
+		if !vfn() {
+			runtime.Abort("State mutated outside of Transaction() scope")
+		}
 	}
+}
 
+// Dragons: move this to the storage API
+func (h *actorStateHandle) cidOf(obj interface{}) cid.Cid {
+	// get Cid for object
+	// Dragons: this should not be storing
```
Comment (anorth) [inline internal/pkg/vm/internal/vmcontext/actor_state_handle.go]: This is easy enough to fix now. Factor the `if` block out of `storage.Put()` into an exported `storage.ToNode(interface{})` and call that.

File: internal/pkg/vm/internal/vmcontext/actor_state_handle_test.go [hunk 6]
```diff
@@ -11,120 +13,241 @@ import (
...
+		var out testActorStateHandleState
+
+		ts.h.Transaction(&out, func() (interface{}, error) {
+			out.FieldA = "changed!"
+			return nil, nil
+		})
+
+		out.FieldA = "changed again!"
 	})
+}
```
Comment (anorth) [inline internal/pkg/vm/internal/vmcontext/actor_state_handle_test.go]: Add a test for sequential transactions, and for readonly followed by transaction.

File: internal/pkg/vm/testing.go (hunk 1 above)
Comment (anorth) [inline internal/pkg/vm/testing.go]: Update the comment on line 230

Comment (icorderi) [reply internal/pkg/vm/testing.go]: `builtin/miner/miner_test.go:184` the `TestConstructor` added a ~week ago was not passing now if didn't add that `nil` check.

File: internal/pkg/vm/internal/vmcontext/actor_state_handle.go [hunk 7]
```diff
@@ -9,62 +9,110 @@ import (
...
 // NewActorStateHandle returns a new `actorStateHandle`
 func NewActorStateHandle(ctx actorStateHandleContext, head cid.Cid) runtime.ActorStateHandle {
-	return &actorStateHandle{
-		ctx:  ctx,
-		head: &head,
+	aux := newActorStateHandle(ctx, head)
+	return &aux
+}
+
+func newActorStateHandle(ctx actorStateHandleContext, head cid.Cid) actorStateHandle {
```
[the remark this answers is not in the document]
Comment (icorderi) [reply internal/pkg/vm/internal/vmcontext/actor_state_handle.go]: resolved offline

**Pull request merged**